{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from unipath import Path\n",
      "from Bio import GenBank\n",
      "import pprint\n",
      "import os\n",
      "import time"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "records = []\n",
      "for p in Path('./flat_files/').listdir():\n",
      "    try:\n",
      "        gbr = GenBank.read(open(p))\n",
      "        records.append(gbr)\n",
      "    except:\n",
      "        print 'error with file', p\n",
      "print \"there are %s records\" % len(records)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "there are 1163 records\n"
       ]
      }
     ],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "schema = {}\n",
      "for record in records:\n",
      "    for key, value in record.__dict__.iteritems():\n",
      "        if key not in schema:\n",
      "            schema[key] = {}\n",
      "    for feature in record.features:\n",
      "        if not feature.key in schema['features']:\n",
      "            schema['features'][feature.key] = set([])\n",
      "        else:\n",
      "            for qualifier in feature.qualifiers:\n",
      "                schema['features'][feature.key].add(qualifier.key)\n",
      "pprint.pprint(schema)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "{'accession': {},\n",
        " 'base_counts': {},\n",
        " 'comment': {},\n",
        " 'contig': {},\n",
        " 'data_file_division': {},\n",
        " 'date': {},\n",
        " 'db_source': {},\n",
        " 'dblinks': {},\n",
        " 'definition': {},\n",
        " 'features': {\"3'UTR\": set(['/allele=',\n",
        "                            '/citation=',\n",
        "                            '/experiment=',\n",
        "                            '/gene=',\n",
        "                            '/gene_synonym=',\n",
        "                            '/inference=',\n",
        "                            '/note=']),\n",
        "              \"5'UTR\": set(['/allele=',\n",
        "                            '/function=',\n",
        "                            '/gene=',\n",
        "                            '/gene_synonym=']),\n",
        "              'CAAT_signal': set(['/note=']),\n",
        "              'CDS': set(['/EC_number=',\n",
        "                          '/allele=',\n",
        "                          '/citation=',\n",
        "                          '/codon_start=',\n",
        "                          '/db_xref=',\n",
        "                          '/exception=',\n",
        "                          '/experiment=',\n",
        "                          '/function=',\n",
        "                          '/gene=',\n",
        "                          '/gene_synonym=',\n",
        "                          '/inference=',\n",
        "                          '/locus_tag=',\n",
        "                          '/note=',\n",
        "                          '/product=',\n",
        "                          '/protein_id=',\n",
        "                          '/pseudo',\n",
        "                          '/standard_name=',\n",
        "                          '/translation=']),\n",
        "              'GC_signal': set(['/gene=']),\n",
        "              'TATA_signal': set(['/gene=']),\n",
        "              'V_segment': set(['/gene=', '/pseudo', '/standard_name=']),\n",
        "              'enhancer': set(['/allele=',\n",
        "                               '/bound_moiety=',\n",
        "                               '/gene=',\n",
        "                               '/gene_synonym=',\n",
        "                               '/note=']),\n",
        "              'exon': set(['/allele=',\n",
        "                           '/citation=',\n",
        "                           '/experiment=',\n",
        "                           '/function=',\n",
        "                           '/gene=',\n",
        "                           '/gene_synonym=',\n",
        "                           '/inference=',\n",
        "                           '/note=',\n",
        "                           '/number=',\n",
        "                           '/product=',\n",
        "                           '/pseudo',\n",
        "                           '/standard_name=']),\n",
        "              'gap': set(['/estimated_length=']),\n",
        "              'gene': set(['/allele=',\n",
        "                           '/experiment=',\n",
        "                           '/gene=',\n",
        "                           '/gene_synonym=',\n",
        "                           '/locus_tag=',\n",
        "                           '/note=',\n",
        "                           '/pseudo']),\n",
        "              'intron': set(['/allele=',\n",
        "                             '/citation=',\n",
        "                             '/experiment=',\n",
        "                             '/function=',\n",
        "                             '/gene=',\n",
        "                             '/gene_synonym=',\n",
        "                             '/note=',\n",
        "                             '/number=',\n",
        "                             '/pseudo']),\n",
        "              'mRNA': set(['/allele=',\n",
        "                           '/experiment=',\n",
        "                           '/gene=',\n",
        "                           '/gene_synonym=',\n",
        "                           '/inference=',\n",
        "                           '/locus_tag=',\n",
        "                           '/note=',\n",
        "                           '/product=',\n",
        "                           '/pseudo']),\n",
        "              'mat_peptide': set(['/allele=',\n",
        "                                  '/experiment=',\n",
        "                                  '/gene=',\n",
        "                                  '/note=',\n",
        "                                  '/product=',\n",
        "                                  '/pseudo']),\n",
        "              'misc_binding': set(['/bound_moiety=',\n",
        "                                   '/citation=',\n",
        "                                   '/experiment=']),\n",
        "              'misc_difference': set(['/citation=',\n",
        "                                      '/gene=',\n",
        "                                      '/note=',\n",
        "                                      '/replace=']),\n",
        "              'misc_feature': set(['/allele=',\n",
        "                                   '/citation=',\n",
        "                                   '/gene=',\n",
        "                                   '/gene_synonym=',\n",
        "                                   '/inference=',\n",
        "                                   '/locus_tag=',\n",
        "                                   '/note=',\n",
        "                                   '/pseudo']),\n",
        "              'misc_recomb': set(['/gene=', '/note=']),\n",
        "              'misc_signal': set(['/function=',\n",
        "                                  '/gene=',\n",
        "                                  '/note=',\n",
        "                                  '/phenotype=']),\n",
        "              'polyA_signal': set(['/gene=',\n",
        "                                   '/inference=',\n",
        "                                   '/locus_tag=',\n",
        "                                   '/note=']),\n",
        "              'polyA_site': set(['/gene=', '/inference=', '/locus_tag=']),\n",
        "              'prim_transcript': set(),\n",
        "              'primer_bind': set(),\n",
        "              'promoter': set(['/experiment=', '/gene=', '/gene_synonym=']),\n",
        "              'protein_bind': set(['/bound_moiety=',\n",
        "                                   '/function=',\n",
        "                                   '/gene=',\n",
        "                                   '/note=']),\n",
        "              'repeat_region': set(['/allele=',\n",
        "                                    '/citation=',\n",
        "                                    '/experiment=',\n",
        "                                    '/gene=',\n",
        "                                    '/note=',\n",
        "                                    '/rpt_family=',\n",
        "                                    '/rpt_type=',\n",
        "                                    '/rpt_unit_range=',\n",
        "                                    '/rpt_unit_seq=',\n",
        "                                    '/satellite=',\n",
        "                                    '/standard_name=']),\n",
        "              'sig_peptide': set(['/experiment=', '/gene=', '/note=']),\n",
        "              'source': set(['/PCR_primers=',\n",
        "                             '/cell_line=',\n",
        "                             '/cell_type=',\n",
        "                             '/chromosome=',\n",
        "                             '/clone=',\n",
        "                             '/clone_lib=',\n",
        "                             '/collected_by=',\n",
        "                             '/collection_date=',\n",
        "                             '/country=',\n",
        "                             '/db_xref=',\n",
        "                             '/dev_stage=',\n",
        "                             '/germline',\n",
        "                             '/haplotype=',\n",
        "                             '/isolate=',\n",
        "                             '/isolation_source=',\n",
        "                             '/lab_host=',\n",
        "                             '/map=',\n",
        "                             '/mol_type=',\n",
        "                             '/note=',\n",
        "                             '/organism=',\n",
        "                             '/sex=',\n",
        "                             '/tissue_type=']),\n",
        "              'unsure': set(['/gene=', '/locus_tag=', '/note=']),\n",
        "              'variation': set(['/allele=',\n",
        "                                '/compare=',\n",
        "                                '/frequency=',\n",
        "                                '/gene=',\n",
        "                                '/note=',\n",
        "                                '/replace='])},\n",
        " 'gi': {},\n",
        " 'keywords': {},\n",
        " 'locus': {},\n",
        " 'nid': {},\n",
        " 'organism': {},\n",
        " 'origin': {},\n",
        " 'pid': {},\n",
        " 'primary': {},\n",
        " 'projects': {},\n",
        " 'references': {},\n",
        " 'residue_type': {},\n",
        " 'segment': {},\n",
        " 'sequence': {},\n",
        " 'size': {},\n",
        " 'source': {},\n",
        " 'taxonomy': {},\n",
        " 'version': {},\n",
        " 'wgs': {},\n",
        " 'wgs_scafld': {}}\n"
       ]
      }
     ],
     "prompt_number": 41
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from sqlalchemy import create_engine\n",
      "from sqlalchemy import Column, Integer, String, Text\n",
      "from sqlalchemy.ext.declarative import declarative_base\n",
      "from sqlalchemy import ForeignKey\n",
      "from sqlalchemy.orm import sessionmaker, relationship, backref\n",
      "\n",
      "SETTINGS = {\n",
      "            'DB':{\n",
      "                  'USER':'root',\n",
      "                  'LOCATION':'localhost',\n",
      "                  'PORT':'3306',\n",
      "                  'DATABASE_NAME':'chromo9',\n",
      "                  },\n",
      "            }\n",
      "\n",
      "class Settings:\n",
      "    def __init__(self, **attrs):\n",
      "        for key, value in attrs.iteritems():\n",
      "            if type(value) is dict:\n",
      "                attrs[key] = Settings(**value)\n",
      "        self.__dict__.update(attrs)\n",
      "        \n",
      "SETTINGS = Settings(**SETTINGS)\n",
      "\n",
      "if not hasattr(SETTINGS.DB, 'PASSWD'):\n",
      "    SETTINGS.DB.PASSWD = raw_input('Enter passwd for mysql user {}'.format(SETTINGS.DB.USER))\n",
      "\n",
      "conn_string = 'mysql+mysqldb://{user}:{passwd}@{location}:{port}'.format(\n",
      "                                            user = SETTINGS.DB.USER,\n",
      "                                            passwd = SETTINGS.DB.PASSWD,\n",
      "                                            location = SETTINGS.DB.LOCATION,\n",
      "                                            port = SETTINGS.DB.PORT,\n",
      "                                                                         )\n",
      "#engine = create_engine('sqlite:///:memory:')\n",
      "engine = create_engine(conn_string, echo=False)\n",
      "engine.execute('CREATE DATABASE IF NOT EXISTS {}'.format(SETTINGS.DB.DATABASE_NAME))\n",
      "engine.execute('USE {}'.format(SETTINGS.DB.DATABASE_NAME))\n",
      "Session = sessionmaker(bind=engine)\n",
      "session = Session()\n",
      "\n",
      "Base = declarative_base()\n",
      "\n",
      "class Gene(Base):\n",
      "    __tablename__ = 'genes'\n",
      "    \n",
      "    gene_identifier= Column(Integer, primary_key=True)\n",
      "    nucleotide_sequence = Column(Text)\n",
      "    chromosome_location = Column(String(8))    \n",
      "\n",
      "#    accession_numbers = relationship(\"Accession\", backref=\"gene\")\n",
      "#    exons = relationship(\"Exon\", backref=\"gene\")\n",
      "#    protein = relationship(\"Protein\", backref=\"gene\")\n",
      "\n",
      "\n",
      "class Accession(Base):\n",
      "    __tablename__ = 'accession_numbers'\n",
      "    \n",
      "    accession_number = Column(String(8), primary_key=True)\n",
      "    gene_id = Column(Integer, ForeignKey('genes.gene_identifier'))\n",
      "    \n",
      "    #gene = relationship(Gene, backref='accession_numbers')\n",
      "\n",
      "\n",
      "class Exon(Base):\n",
      "    __tablename__ = 'exons'\n",
      "    \n",
      "    id = Column(Integer, primary_key=True)\n",
      "    start = Column(Integer)\n",
      "    end = Column(Integer)\n",
      "    gene_id = Column(Integer, ForeignKey('genes.gene_identifier'))\n",
      "    \n",
      "    #gene = relationship(Gene, backref='exons')\n",
      "\n",
      "\n",
      "class Protein(Base):\n",
      "    __tablename__ = 'proteins'\n",
      "\n",
      "    id = Column(Integer, primary_key=True)    \n",
      "    name = Column(String(128))\n",
      "    sequence = Column(Text)\n",
      "    gene_id = Column(Integer, ForeignKey('genes.gene_identifier'))\n",
      "\n",
      "    #gene = relationship(Gene, backref='protein_products')\n",
      "\n",
      "Base.metadata.create_all(engine) # create the tables"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Enter passwd for mysql user rootsn0wb1rd\n"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "session.rollback()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "for record in records:\n",
      "    gene = Gene(gene_identifier=int(record.gi) ,nucleotide_sequence=record.sequence ,chromosome_location=record.locus)\n",
      "    session.add(gene)\n",
      "    session.commit()\n",
      "\n",
      "    for feature in record.features:\n",
      "        \n",
      "        if feature.key == 'CDS':\n",
      "            translation, name  = None, None\n",
      "            for qualifier in feature.qualifiers:\n",
      "                if qualifier.key == '/translation=':\n",
      "                    translation = qualifier.value.strip('\"')\n",
      "                if qualifier.key == '/product=':\n",
      "                    name = qualifier.value.strip('\"')\n",
      "            # create protein object\n",
      "            protein = Protein(name=name, sequence=translation, gene_id=gene.gene_identifier)\n",
      "            session.add(protein)\n",
      "            session.commit()\n",
      "            \n",
      "        if feature.key == 'exon':\n",
      "            start, end = feature.location.split('..')\n",
      "            start, end = int(''.join([d for d in start if d.isdigit()])), int(''.join([d for d in end if d.isdigit()]))\n",
      "            exon = Exon(start=start,end=end,gene_id=gene.gene_identifier)\n",
      "            session.add(exon)\n",
      "            session.commit()\n",
      "\n",
      "    for accession in record.accession:\n",
      "        acc = Accession(accession_number=accession, gene_id=gene.gene_identifier)\n",
      "        session.add(acc)\n",
      "        session.commit()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/default.py:324: Warning: Data truncated for column 'nucleotide_sequence' at row 1\n",
        "  cursor.execute(statement, parameters)\n",
        "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/default.py:324: Warning: Data truncated for column 'chromosome_location' at row 1\n",
        "  cursor.execute(statement, parameters)\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/default.py:324: Warning: Data truncated for column 'name' at row 1\n",
        "  cursor.execute(statement, parameters)\n"
       ]
      }
     ],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from collections import Counter\n",
      "\n",
      "bases = ['t', 'c', 'a', 'g']\n",
      "codons = [a+b+c for a in bases for b in bases for c in bases]\n",
      "amino_acids = 'FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG'\n",
      "codon_table = dict(zip(codons, amino_acids))\n",
      "\n",
      "def translate(seq):\n",
      "    seq = seq.lower().replace('\\n', '').replace(' ', '')\n",
      "    peptide = ''\n",
      "    \n",
      "    for i in xrange(0, len(seq), 3):\n",
      "        codon = seq[i: i+3]\n",
      "        amino_acid = codon_table.get(codon, '*')\n",
      "        if amino_acid != '*':\n",
      "            peptide += amino_acid\n",
      "        else:\n",
      "            break\n",
      "                \n",
      "    return peptide\n",
      "\n",
      "def transcribe(seq):\n",
      "    mapping = {'A':'U', 'T':'A', 'C':'G', 'G':'U'}\n",
      "    try:\n",
      "        cdna = ''.join([mapping[base] for base in seq])\n",
      "        return cdna\n",
      "    except:\n",
      "        print \"\\nError\\n\", seq\n",
      "\n",
      "def codon_frequencies(seq):\n",
      "    assert seq % 3 == 0\n",
      "    return Counter([seq[i:i+3] for i in xrange(0,len(seq),3)])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 51
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "\n",
      "timestamp = time.strftime('%Y-%m-%d-%I:%M')\n",
      "dumpstring = \"mysqldump -u {user} -p{passwd} -h {location} -e --opt -c {database} | gzip -c > {database}_{timestamp}.gz\".format(\n",
      "                    user = SETTINGS.DB.USER,\n",
      "                    passwd = SETTINGS.DB.PASSWD,\n",
      "                    location = SETTINGS.DB.LOCATION,\n",
      "                    database = SETTINGS.DB.DATABASE_NAME,\n",
      "                    timestamp = timestamp,\n",
      "                    )\n",
      "os.popen(dumpstring)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 10,
       "text": [
        "<open file 'mysqldump -u root -psn0wb1rd -h localhost -e --opt -c chromo9 | gzip -c > chromo9_2014-03-12-05:30.gz', mode 'r' at 0xc70a150>"
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from sqlalchemy import create_engine\n",
      "from sqlalchemy import Column, Integer, String, Text\n",
      "from sqlalchemy.ext.declarative import declarative_base\n",
      "from sqlalchemy import ForeignKey\n",
      "from sqlalchemy.orm import sessionmaker, relationship, backref\n",
      "\n",
      "class Models():\n",
      "    def __init__(self):\n",
      "        Base = declarative_base()\n",
      "        self.Base = Base\n",
      "        \n",
      "        class Gene(Base):\n",
      "            __tablename__ = 'genes'\n",
      "            \n",
      "            gene_identifier= Column(Integer, primary_key=True)\n",
      "            nucleotide_sequence = Column(Text)\n",
      "            chromosome_location = Column(String(8))    \n",
      "        \n",
      "        #    accession_numbers = relationship(\"Accession\", backref=\"gene\")\n",
      "        #    exons = relationship(\"Exon\", backref=\"gene\")\n",
      "        #    protein = relationship(\"Protein\", backref=\"gene\")\n",
      "        \n",
      "        \n",
      "        class Accession(Base):\n",
      "            __tablename__ = 'accession_numbers'\n",
      "            \n",
      "            accession_number = Column(String(8), primary_key=True)\n",
      "            gene_id = Column(Integer, ForeignKey('genes.gene_identifier'))\n",
      "            \n",
      "            #gene = relationship(Gene, backref='accession_numbers')\n",
      "        \n",
      "        \n",
      "        class Exon(Base):\n",
      "            __tablename__ = 'exons'\n",
      "            \n",
      "            id = Column(Integer, primary_key=True)\n",
      "            start = Column(Integer)\n",
      "            end = Column(Integer)\n",
      "            gene_id = Column(Integer, ForeignKey('genes.gene_identifier'))\n",
      "            \n",
      "            #gene = relationship(Gene, backref='exons')\n",
      "        \n",
      "        \n",
      "        class Protein(Base):\n",
      "            __tablename__ = 'proteins'\n",
      "        \n",
      "            id = Column(Integer, primary_key=True)    \n",
      "            name = Column(String(128))\n",
      "            sequence = Column(Text)\n",
      "            gene_id = Column(Integer, ForeignKey('genes.gene_identifier'))\n",
      "        \n",
      "            #gene = relationship(Gene, backref='protein_products')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "mod = Models()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "mod.Base.metadata.create_all(engine) # create the tables"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "NameError",
       "evalue": "name 'engine' is not defined",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
        "\u001b[1;32m<ipython-input-8-1f0945f2c050>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mmod\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mBase\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmetadata\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcreate_all\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mengine\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m# create the tables\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
        "\u001b[1;31mNameError\u001b[0m: name 'engine' is not defined"
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}